Allow multi-axis `roll` as in numpy 1.12+

grlee77 opened the pull request:
For ease of review, the body of `numpy.roll` function is reproduced exactly in the first commit here.
The minimal changes needed for `cupy` are in the second commit.

grlee77 commented:
Importing `normalize_axis_tuple` from older numpy fails here. This small function could be copied locally to fix the issue.

niboshi commented:
Thank you for PR!
Could you add more tests?
What I could come up with are:

* More axes than ndim
* `shift` is a 2-element tuple and `axis` is a scalar
* `shift` is a tuple and `axis` is `None`
* There are unspecified axes in `axis`, e.g. `ndim=3` and `axis=(1,2)`
* Negative axis that turns out to duplicate, e.g. `ndim=2` and `axis=(-1, 1)`.

grlee77 commented:
I have added several more tests and think I have covered all the cases you mentioned above.

We can remove `accept_error` from the tests using `numpy_cupy_raises` if it is not desired to check that the error type is the same.

niboshi commented on cupy/manipulation/rearrange.py:
```diff
@@ -1,6 +1,76 @@
 from cupy import core
 
+try:
+    from numpy.core.numeric import normalize_axis_tuple
```
We don't think we should use `numpy.core` module, because it's an undocumented module.
We should avoid using copy-pasted NumPy code either, because the feature is simple enough to be implemented by ourselves.

Could you either re-implement it, or use (or extend) the existing normalization logic?
The existing implementation is below:
https://github.com/cupy/cupy/blob/0fe2fd9a4ab1d7a1f14aceba780ca59828dd8c75/cupy/core/reduction.pxi#L91)

grlee77 replied:
I found that in numpy <=1.12 an out-of-bounds axis returns `ValueError`, but starting from 1.13, it returns a newly defined `numpy.AxisError` instead (where `numpy.AxisError` inherits from both `ValueError` and `IndexError`).  That is why the test cases with `accept_error='numpy.AxisError'` specify numpy>=1.13. 

Aside from changing the issue with the error type, the existing `_get_axis` logic should work here.

niboshi replied:
Ah, we have `core.core._AxisError` for that reason I think. How about raising it?
That way I think `accept_error` can be omitted.

grlee77 replied:
Okay, done.  I guess I should have grepped for `AxisError` first.  I just saw that `cupy.AxisError` did not exist (perhaps it should, but that is separate from the issue in this PR)

niboshi commented on tests/cupy_tests/manipulation_tests/test_rearrange.py:
```diff
@@ -1,5 +1,6 @@
 import unittest
 
+import numpy
```
Please add a blank line after `import numpy`.
We group imports in three categories: standard libraries, third-party libraries, internal imports.
(https://docs.chainer.org/en/v6.0.0b1/contribution.html#coding-guidelines)

grlee77 replied:
done

grlee77 commented:
I removed the copied `normalize_axis_tuple` and added a new test case verifying that duplicates in `axis` are allowed.

please check if the change to `_get_axis` error handling is acceptable.

grlee77 commented:
I have fixed the Travis error. Thank you for the helpful review!

niboshi commented:
Thanks!
Jenkins, test this please.

niboshi commented:
It seems `numpy.broadcast.ndim` was introduced in NumPy 1.12.0.
How about using `nd` instead? (it's not recommended in documentation but not deprecated.)

https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.broadcast.html#numpy.broadcast

niboshi commented:
Jenkins, test this please

niboshi commented:
LGTM.
Thank you very much!
